### PR TITLE
test: add guardrail e2e coverage (presidio masking, bedrock post and during call, moderation on messages)

### PR DESCRIPTION
**Pull request by yucheng-berri**
## TLDR

Problem this solves:

- Guardrail e2e coverage missed presidio masking, bedrock post_call/during_call, and /v1/messages moderation
- No e2e proved a streamed request is blocked before the first chunk

How it solves it:

- presidio pre_call masking proven live on /v1/chat/completions and /v1/messages
- bedrock post_call proven to block denied MODEL OUTPUT without leaking it
- bedrock during_call + stream=true proven to block as a JSON 400 with zero chunks delivered
- openai moderation block/pass proven through /v1/messages

## User Flow

Before: a proxy admin enabling PII masking or streaming guardrails has no e2e telling them enforcement broke

1. They register a presidio guardrail via POST /guardrails and send PII through POST /v1/chat/completions
2. Nothing in the suite asserts the model only ever saw `<EMAIL_ADDRESS>` placeholders
3. A streamed request that violates a bedrock policy has no proof it is stopped before content reaches the client

After: each enforcement point has a live assertion

1. The suite registers presidio (pre_call, input scope) per test and proves the echoed reply carries placeholders and never the raw email/phone, on both /v1/chat/completions and /v1/messages
2. It registers bedrock post_call, makes the model emit the word-policy-denied word, and expects the block without the output leaking into the error body
3. It registers bedrock during_call, sends `stream=true`, and expects HTTP 400 JSON with zero delivered chunks and no event-stream content type
4. It sends a flagged and a benign prompt through POST /v1/messages behind openai moderation and expects a 400 naming moderation, then a normal answer

## Relevant issues

N/A - coverage work, no tracking issue.

## Pre-Submission checklist

- [x] I have added meaningful tests
- [x] The handful of test files covering my change pass locally, e.g. `uv run pytest tests/test_litellm/<your_test_file>.py -v`. Leave the suites (`make test-unit-*`, `make test-unit`) to CI: it finishes in ~15 minutes where a laptop takes an hour or more
- [x] My PR passes all required CI/CD checks (e.g., lint, schema.d.ts sync check, etc.)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [ ] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Screenshots / Proof of Fix

Setup: live proxy (this branch) on port 4620 with real Postgres/Redis, real Presidio analyzer+anonymizer containers, real Bedrock guardrail (word policy denying FORBIDDENWORD) signed with real AWS credentials, real OpenAI moderation, gemini-2.5-flash upstream.

### Before (bcc2e71b5d)

1. The four registry cells (`guardrail.presidio.pre_call.masks`, `guardrail.bedrock.post_call.blocks`, `guardrail.bedrock.during.blocks`, plus `guardrail.openai_moderations.pre_call.blocks` on /v1/messages) have no claiming test on their scenarios.

### After (e6a9bec6a1)

1. `PYTHONPATH=. pytest guardrails/test_presidio_masking_e2e.py guardrails/test_streaming_guardrail_e2e.py guardrails/test_bedrock_guardrail_e2e.py guardrails/test_openai_moderation_guardrail_e2e.py -q`
2. Observed: all 7 scenarios green against the live proxy - masked echoes carry `EMAIL_ADDRESS` and never the raw PII; the during_call streamed block is a non-SSE 400 with `chunks == 0`; the post_call block body names the guardrail without the denied word; moderation blocks the flagged /v1/messages prompt with a 400 naming moderation and passes the benign one
3. After the review-driven hardening (stream poll retries through transient -1/401/429 statuses, the masking loop retries non-Success results, the vacuous post-emptiness leak assertion dropped), the three touched tests were rerun live: `3 passed in 46.86s`
4. The masking predicate now also requires the `PHONE_NUMBER` placeholder positively; both presidio tests rerun live: `2 passed in 31.69s`
5. After the Bugbot round the post_call predicate accepts only the 400 verdict body (violated/blocked/intervened), so a failed ApplyGuardrail 403 cannot pass as a block; rerun live at this head: `1 passed in 16.34s`

## Type

✅ Test

### Medium

- The buildkite e2e check on this PR is green but vacuous: the full suite only runs for PRs based on litellm_internal_staging, and this PR is stacked on the logging branch
  - on the cluster the bedrock tests also need BerriAI/litellm-ops#156, which wires BEDROCK_GUARDRAIL_BLOCKED_WORD (the stage guardrail denies "bread", not the test default FORBIDDENWORD)
- A presidio guardrail registered via POST /guardrails with `mode="logging_only"` does NOT mask the shipped payload today: the standard logging payload is built before `async_logging_hook` runs, so s3/gcs/datadog receive the raw prompt (live-proven with a marker probe reading the S3 object back)
  - the planned `guardrail.presidio.logging_only.masks` test was therefore dropped from this PR; the cell stays uncovered until the product gap is fixed

### Low

- The presidio and bedrock post_call tests depend on the model echoing text verbatim; token budgets are sized (max_tokens=128) so reasoning tokens cannot starve the echo
- `poll_until_blocked` is generified to any pydantic result so /v1/messages outcomes poll too; existing callers are source-compatible

## QA runbook

- tests/e2e/guardrails/test_presidio_masking_e2e.py::TestPresidioPreCallMasking::test_pre_call_masks_pii_on_chat_completions - presidio pre_call masks PII before the model sees it
  - [ ] POST /guardrails with `{"guardrail": "presidio", "mode": "pre_call", "presidio_filter_scope": "input", analyzer/anonymizer bases}` (needs running presidio containers)
  - [ ] POST /v1/chat/completions asking the model to repeat a fake email + 555 phone, `guardrails: [<name>]`
  - [ ] Expect the reply to echo both `EMAIL_ADDRESS` and `PHONE_NUMBER` placeholders and contain neither the raw email nor the phone (retry to ~40s for guardrail propagation)
  - [ ] Sanity check: this test makes sense to add and is not hand-wavey (unmasked early responses are treated as propagation, only a raw echo at the deadline fails)
- tests/e2e/guardrails/test_presidio_masking_e2e.py::TestPresidioPreCallMasking::test_pre_call_masks_pii_on_messages - the same contract through POST /v1/messages
  - [ ] Same guardrail; POST /v1/messages with the PII prompt
  - [ ] Expect placeholder-echo and no raw PII in the answer text
  - [ ] Sanity check: this test makes sense to add and is not hand-wavey
- tests/e2e/guardrails/test_bedrock_guardrail_e2e.py::TestBedrockGuardrail::test_bedrock_post_call_blocks_denied_model_output - bedrock post_call blocks denied model OUTPUT
  - [ ] POST /guardrails with `{"guardrail": "bedrock", "mode": "post_call", guardrailIdentifier/Version}` (word policy must deny FORBIDDENWORD, override via BEDROCK_GUARDRAIL_BLOCKED_WORD)
  - [ ] POST /v1/chat/completions: "Reply with exactly this one word and nothing else: FORBIDDENWORD"
  - [ ] Expect 400/403 whose body names the guardrail and does NOT contain the denied word
  - [ ] Sanity check: this test makes sense to add and is not hand-wavey (the denied word must come out of the model - the prompt itself is not scanned in post_call)
- tests/e2e/guardrails/test_streaming_guardrail_e2e.py::TestBedrockDuringCallStreaming::test_during_call_blocks_stream_before_first_chunk - a during_call block beats the stream
  - [ ] POST /guardrails with mode during_call; POST /v1/chat/completions with `stream: true` and the denied word in the prompt
  - [ ] Expect HTTP 400 JSON (not text/event-stream) with zero SSE chunks delivered
  - [ ] Sanity check: this test makes sense to add and is not hand-wavey
- tests/e2e/guardrails/test_openai_moderation_guardrail_e2e.py::TestOpenAIModerationGuardrail::test_moderation_blocks_flagged_input_on_messages - moderation gates /v1/messages
  - [ ] POST /guardrails with `{"guardrail": "openai_moderation", "mode": "pre_call"}`; POST /v1/messages with a flagged prompt then a benign one
  - [ ] Expect a 400 naming moderation, then a normal answer for the benign prompt
  - [ ] Sanity check: this test makes sense to add and is not hand-wavey

## Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

### CODE REVIEW

**Comment by yucheng-berri** [comment]
bugbot run

File: tests/e2e/guardrails/test_presidio_masking_e2e.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+from guardrails_client import GuardrailsClient, PresidioParamsBody
+from lifecycle import ResourceManager
+from models import AnthropicMessagesResponse, ChatResponse
+
+pytestmark = pytest.mark.e2e
+
+MODEL = "gemini-2.5-flash"
+
+# A guardrail created via POST /guardrails reaches the worker that served the
+# create immediately, but every other worker only picks it up on its next
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply tests/e2e/guardrails/test_presidio_masking_e2e.py:36]: This follows the suite pattern: the pre-existing bedrock guardrail test pins the same module-level MODEL, served by the shared gateway config

File: tests/e2e/guardrails/test_presidio_masking_e2e.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    while True:
+        result = fetch()
+        match result:
+            case Success(data=data):
+                content = extract(data)
+                last = content
+                if MASKED_EMAIL_TOKEN in content and email not in content:
+                    assert FAKE_PHONE not in content, (
+                        f"the raw phone number must be masked before the model sees it, but the "
+                        f"response echoed it: {content[:300]!r}"
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply tests/e2e/guardrails/test_presidio_masking_e2e.py:134]: Fixed in 4dbc907813, the predicate now requires the PHONE_NUMBER placeholder positively, and both presidio tests pass again on the live rig

**Comment by yucheng-berri** [comment]
@greptileai

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai

**Comment by yucheng-berri** [comment]
bugbot run

File: tests/e2e/guardrails/guardrails_client.py [hunk 3]
```diff
@@ -337,3 +377,25 @@ def poll_until_blocked(call: Callable[[], Result[ChatResponse]]) -> Result[ChatR
...
+    """poll_until_blocked for raw/streamed sends, which return a StreamingResponse
+    instead of a Result: retry while the call still succeeds (the data-plane worker
+    has not picked the new guardrail up yet) or fails with a transient status,
+    returning the first guardrail-shaped non-2xx outcome or the last result at
+    the deadline."""
+    deadline = time.monotonic() + POLL_TIMEOUT
+    last = call()
+    while time.monotonic() < deadline:
+        if not last.ok and last.status_code not in _TRANSIENT_STREAM_STATUSES:
+            return last
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply tests/e2e/guardrails/guardrails_client.py:398]: Probed live: a request naming an unsynced guardrail returns 200 on stream and non-stream, the proxy ignores it, so no 404 propagation shape exists

File: tests/e2e/guardrails/test_bedrock_guardrail_e2e.py [hunk 4]
```diff
@@ -75,3 +82,52 @@ def test_bedrock_pre_call_blocks_harmful_prompt(
...
+        match result:
+            case UnknownApiError(status_code=status, body=body):
+                assert status in {400, 403}, (
+                    f"expected a guardrail block status, got {status}: {body[:400]}"
+                )
+                body_lower = body.lower()
+                assert any(
+                    token in body_lower
+                    for token in ("guardrail", "violated", "blocked", "intervened")
+                ), f"block body should name the guardrail reason; got: {body[:400]}"
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply tests/e2e/guardrails/test_bedrock_guardrail_e2e.py]: Fixed in b5f994ed87, the predicate now requires the 400 verdict body naming violated, blocked, or intervened, and the test passes again live

**Comment by yucheng-berri** [comment]
@greptileai

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai

**Comment by yucheng-berri** [comment]
bugbot run
